core/src/main/resources/jenkins/console/ConsoleUrlProviderUserProperty/config.jelly

```diff
@@ -0,0 +1,15 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <j:if test="${descriptor.enabled}">
+        <f:invisibleEntry>
+          <!-- Dummy field to always have an entry in form submissions so users can delete all providers. -->
```

Comment:
Do not do it this way. Rather start `doConfigSubmit` or whatever by nulling out the field before binding properties.

Reply:
IIRC, both were needed, otherwise `configure` didn't even get called. I will check again though, perhaps this can be deleted.

Reply:
Ah yeah I think this behavior was specific to user properties actually. I'll still check it again though.

Reply:
Probably what you ran into is that form binding only provides a JSON element for a list if it has at least one entry. So for a top-level (reconfigurable) object, if you just use the default `bindJSON`, after deleting all elements of a list and saving, no change is applied to that property. The workaround is to simply null out any list-valued fields before calling `bindJSON`.

Reply:
For `ConsoleUrlProviderGlobalConfiguration`, yes, that was a problem and that is how I fixed it.

`ConsoleUrlProviderUserProperty` is different though. [`User.doConfigSubmit`](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L867) is unusual and has complex logic regarding creating new properties vs reconfiguring existing ones. IIRC, in the problematic case, since `ConsoleUrlProviderUserProperty` only has a single `List` field, [this condition](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881) is `true` because the property itself is not present in the submitted JSON at all, and so the logic there chooses to reuse the existing property if one exists.

Whether that logic was intentional was not clear to me. Perhaps [these lines](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L889-L892) should be moved inside of [this block](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881-L887), so that any properties not present in the submission are deleted, but I have no idea if anything relies on the current behavior (perhaps this case is intended to preserve user properties that do not offer any GUI configuration?) and it was not something I wanted to look into here.

Reply:
Hmm, maybe actually I can reuse the lock inside of `TimingFlowNodeStorage` instead to simplify things.

Reply:
> perhaps this case is intended to preserve user properties that do not offer any GUI configuration?

Perhaps. (Or those whose GUI configuration is on a separate page.)

> https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881 is `true` because the property itself is not present in the submitted JSON at all

`false` you mean? Smells like a bug in the form submission JSON generator; it should at least create `{}` maybe with `_class`.

Reply:
Yes sorry I meant `false`.